### PR TITLE
samples: bluetooth: peripheral_proximity

### PR DESCRIPTION
**Pull request by guangli2015**
new add peripheral_proximity sample

subsys: bluetooth: services

add Link loss service and Immediate Alert Service

include: bluetooth: services

add Link loss service and Immediate Alert Service head file
Signed-off-by: andrew li <andrew.li@nordicsemi.no>

### CODE REVIEW

**Comment by maje-emb** [comment]
I think we should add documentation to the service and a changelog entry.
It would be good to split one big commit into several smaller ones.
i.e. one with service implementation, one with sample, one with changelog changes.

**Comment by carlescufi** [comment]
@KAGA164 @alwa-nordic @jori-nordic can you please take a look?

File: samples/bluetooth/peripheral_proximity/prj_minimal.conf [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+CONFIG_MPSL_SIGNAL_STACK_SIZE=640
+CONFIG_MAIN_STACK_SIZE=1024
+CONFIG_IDLE_STACK_SIZE=128
+CONFIG_ISR_STACK_SIZE=1024
+
+# Disable features not needed
+CONFIG_TIMESLICING=n
+CONFIG_MINIMAL_LIBC_MALLOC=n
+CONFIG_LOG=n
+CONFIG_ASSERT=n
```
Comment (jori-nordic) [inline samples/bluetooth/peripheral_proximity/prj_minimal.conf:89]: I know it's also disabled in the `prj_minimal.conf` of the peripheral_hr sample, but I wouldn't disable asserts, as some stack functions depend on it for parameter checking. At the very least, I would leave them in, but commented out, so the user knows it's possible to reduce size at the expense of safety. e.g.:
```
# Disabling asserts might cause undefined behavior,
# only disable if truly necessary
# CONFIG_ASSERT=n
# CONFIG_BT_ASSERT=n
```

File: samples/bluetooth/peripheral_proximity/src/main.c [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+#include <bluetooth/uuid.h>
+#include <bluetooth/gatt.h>
+
+#include <bluetooth/services/ias.h>
+#include <bluetooth/services/lls.h>
+#include <settings/settings.h>
+
+#include <dk_buttons_and_leds.h>
+
+#define DEVICE_NAME             CONFIG_BT_DEVICE_NAME
```
Comment (jori-nordic) [inline samples/bluetooth/peripheral_proximity/src/main.c:29]: unnecessary define

File: samples/bluetooth/peripheral_proximity/src/main.c [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	printk("Disconnected (reason %u)\n", reason);
+
+	dk_set_led_off(CON_STATUS_LED);
+	uint8_t alert_lv = bt_lls_alert_level_get();
+	alert_signal(alert_lv);
+}
+
+
```
Comment (jori-nordic) [inline samples/bluetooth/peripheral_proximity/src/main.c:79]: only one newline between functions. Check the other files too

File: samples/bluetooth/peripheral_proximity/src/main.c [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+static void alert_signal(uint8_t alert_level)
+{
+
+
+	switch (alert_level) {
+	case BLE_CHAR_ALERT_LEVEL_NO_ALERT:
+
+		run_led_blink_off = true;
+		dk_set_led_off(RUN_STATUS_LED);
+		break; // BLE_CHAR_ALERT_LEVEL_NO_ALERT
```
Comment (jori-nordic) [inline samples/bluetooth/peripheral_proximity/src/main.c:103]: please use `/* This style of comment */` instead

File: samples/bluetooth/peripheral_proximity/src/main.c [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+static struct bt_conn_cb conn_callbacks = {
+	.connected = connected,
+	.disconnected = disconnected,
+};
+
+
+static struct bt_conn_auth_cb conn_auth_callbacks;
+
+
+/**@brief Function for the Signals alert event from Immediate Alert or Link Loss services.
```
Comment (jori-nordic) [inline samples/bluetooth/peripheral_proximity/src/main.c:90]: I propose `Set RUN LED based on alert level`.

File: samples/bluetooth/peripheral_proximity/src/main.c [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+	case BLE_CHAR_ALERT_LEVEL_MILD_ALERT:
+
+		run_led_blink_off = false;
+		run_led_blink_interal = RUN_LED_BLINK_MILD_ALERT_INTERVAL;
+
+		break; // BLE_CHAR_ALERT_LEVEL_MILD_ALERT
+
+	case BLE_CHAR_ALERT_LEVEL_HIGH_ALERT:
+
```
Comment (jori-nordic) [inline samples/bluetooth/peripheral_proximity/src/main.c:113]: no newline after the `case`. You can keep the one after the `break`.

File: subsys/bluetooth/services/Kconfig.lls [hunk 7]
```diff
@@ -0,0 +1,19 @@
+#
+# Copyright (c) 2021 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+menuconfig BT_NRF_LLS
+	bool "link lost Service"
```
Comment (jori-nordic) [inline subsys/bluetooth/services/Kconfig.lls:8]: Link Lost

File: subsys/bluetooth/services/lls.c [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+					      NULL, write_alert, NULL),
+		       );
+
+int bt_lls_init()
+{
+
+	alert_level = 0;
+	return 0;
+}
+uint8_t bt_lls_alert_level_get()
```
Comment (jori-nordic) [inline subsys/bluetooth/services/lls.c:72]: newline

File: subsys/bluetooth/services/lls.c [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+					      BT_GATT_CHRC_WRITE,
+					      BT_GATT_PERM_WRITE,
+					      NULL, write_alert, NULL),
+		       );
+
+int bt_lls_init()
+{
+
+	alert_level = 0;
+	return 0;
```
Comment (jori-nordic) [inline subsys/bluetooth/services/lls.c:70]:
```c
alert_level = 0;

return 0;
```

File: subsys/bluetooth/services/lls.c [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+#include <bluetooth/uuid.h>
+#include <bluetooth/gatt.h>
+
+#include <bluetooth/services/lls.h>
+
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(bt_nrf_lls, CONFIG_BT_NRF_LLS_LOG_LEVEL);
+
+static volatile uint8_t alert_level = 0;
```
Comment (jori-nordic) [inline subsys/bluetooth/services/lls.c:26]: Isn't this LLS per-connection ? I mean, shouldn't we have one `alert_level` for each (possible) connection ?

File: subsys/bluetooth/services/ias.c [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+#include <bluetooth/uuid.h>
+#include <bluetooth/gatt.h>
+
+#include <bluetooth/services/ias.h>
+
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(bt_nrf_ias, CONFIG_BT_NRF_IAS_LOG_LEVEL);
+typedef void (*ble_ias_evt_handler_t) (uint8_t para);
+static volatile uint8_t alert_level = 0;
```
Comment (jori-nordic) [inline subsys/bluetooth/services/ias.c:27]: see my comment on L26 of lls.c

File: subsys/bluetooth/services/ias.c [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+#include <bluetooth/conn.h>
+#include <bluetooth/uuid.h>
+#include <bluetooth/gatt.h>
+
+#include <bluetooth/services/ias.h>
+
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(bt_nrf_ias, CONFIG_BT_NRF_IAS_LOG_LEVEL);
+typedef void (*ble_ias_evt_handler_t) (uint8_t para);
```
Comment (jori-nordic) [inline subsys/bluetooth/services/ias.c:26]: did you mean `param` ?

File: subsys/bluetooth/services/ias.c [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+static ssize_t write_alert(struct bt_conn *conn,
+			   const struct bt_gatt_attr *attr,
+			   const void *buf,
+			   uint16_t len, uint16_t offset, uint8_t flags)
+{
+	LOG_DBG("Attribute write, handle: %u, conn: %p", attr->handle,
+		(void *)conn);
+
+	if (len != 1U) {
+		LOG_DBG("Write led: Incorrect data length");
```
Comment (jori-nordic) [inline subsys/bluetooth/services/ias.c:41]: does the service have the concept of LEDs ? or is that just how the callback shows it to the user in the sample ?

File: samples/bluetooth/peripheral_proximity/Kconfig [hunk 14]
```diff
@@ -0,0 +1,11 @@
+#
+# Copyright (c) 2022 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+source "Kconfig.zephyr"
+
+menu "Nordic proximity sample"
```
Comment (jori-nordic) [inline samples/bluetooth/peripheral_proximity/Kconfig:9]: I don't think we need this file, we don't seem to be declaring new sample-specific options

**Review by jori-nordic — COMMENTED** [review]
Some rework needed, I don't see any dependencies on NCS-specific libraries.
Could you please open a PR [upstream](https://github.com/zephyrproject-rtos/zephyr/pulls) instead @guangli2015 ?

**Pull request closed without merging**